Modules/VMware.Skyline.InsightsApi/VMware.Skyline.InsightsApi.psm1

```diff
@@ -0,0 +1,397 @@
+    Created by:	Brian Wuchner
+    Date:		February 21, 2022
+    Blog:		www.enterpriseadmins.org
+    Twitter:		@bwuch
+    ===========================================================================
+    .SYNOPSIS
+    Use this function to post a query to the Skyline Insights API.
+    .DESCRIPTION
+    This function will allow you to query the Skyline Insights API.
+    Proper headers will be formatted and posted if a DefaultSkylineConnection is present.
+    This is primarily a helper function used by other functions included in the module.  
+    It is exported in the module manifest to be used for any custom queries.
+    .EXAMPLE
+    PS C:\> Invoke-SkylineInsightsApi -queryBody '{formatted-query-string-converted-to-json}'
+#>
+    param(
+        [Parameter(Mandatory=$true)][string]$queryBody
+    )
+    
+    if ( !$global:DefaultSkylineConnection ) {
+        Write-Error 'You are not currently connected to any servers. Please connect first using Connect-SkylineInsights.'
+        return;
+    }
+
+    write-debug "Querybody: $queryBody"
+    try {
+        $restCall = invoke-restmethod -method post -Uri $($global:DefaultSkylineConnection.SkylineAPI) -Headers @{Authorization = "Bearer $($global:DefaultSkylineConnection.ConnectionDetail.access_token)"} -body $queryBody -ContentType "application/json"
+        if ($restCall.errors) {
+            Write-Error $restCall.errors.Message
+        }
+        return $restCall
+    } catch {
+        $incomingError = $_
+        try {
+            # are nested try/catch blocks the powershell equilivent of vbscript On Error Resume Next?
+            $errorStatusAsJson = ($incomingError | ConvertFrom-Json).status
+            if ($errorStatusAsJson -eq '429 TOO_MANY_REQUESTS') {
+                write-debug 'Hit 429 / rate limite warning, will sleep for 1 second.'
+                start-sleep -Milliseconds 1005
+                Invoke-SkylineInsightsApi $queryBody
+            }
+        } catch {
+            # this was the error from trying to cast the incoming error to Json
+        }
+        if (!$errorStatusAsJson) { write-error $_ }
+    }
+}
+
+
+Function Get-SkylineFinding {
+<#
+    .NOTES
+    ===========================================================================
+    Created by:	Brian Wuchner
+    Date:		February 21, 2022
+    Blog:		www.enterpriseadmins.org
+    Twitter:		@bwuch
+    ===========================================================================
+    .SYNOPSIS
+    Use this function to query findings from the Skyline Insights API.
+    .DESCRIPTION
+    This function will allow you to query the Skyline Insights API for Findings.
+    As described in the documentation, the maximum limit per page is 200 records.  This function provides
+    an optional pagesize parameter to request smaller batches, but by default assumes 200 records.
+    .EXAMPLE
+    PS C:\> Get-SkylineFinding
+#>
+    [cmdletbinding()]
+    param(
+        [ValidateRange(1,200)][int]$pagesize=200
```

Comment:
The way the cmdlet logic is currently written it will return all the results at the end of the execution. Why would someone want a smaller page size in this case. Either remove this or (preferred solution) change the output not to use return and to actually return the results page by page (https://github.com/PoshCode/PowerShellPracticeAndStyle/issues/46)

Reply:
This parameter only makes sense if you return the results page by page. If not the cmdlet logic will wait until the last item is returned anyway and it will make no difference if you return them in batches of 50 or 200. Normally every new API request introduces a performance penalty and we would like to have as few requests as possible and respectively the largest possible page size.

Reply:
I've set these pagesize parameters to DontShow, which leaves them around for testing small batch queries and forcing many API calls.  I have changed the function to return each batch instead of waiting for completion and returning all results at once.  This will be available in the next commit.